exporter/opentelemetry-exporter-prometheus/src/opentelemetry/exporter/prometheus/__init__.py

```diff
@@ -0,0 +1,218 @@
+    labels = {"environment": "staging"}
+
+    counter.add(25, labels)
+    input("Press any key to exit...")
+
+API
+---
+"""
+
+import collections
+import logging
+import re
+from typing import Iterable, Optional, Sequence, Tuple
+
+from prometheus_client import core
+
+from opentelemetry.context import (
+    _SUPPRESS_INSTRUMENTATION_KEY,
+    attach,
+    detach,
+    set_value,
+)
+from opentelemetry.sdk._metrics.export import MetricReader
+from opentelemetry.sdk._metrics.point import Gauge, Histogram, Metric, Sum
+
+_logger = logging.getLogger(__name__)
+
+
+def _convert_buckets(metric: Metric) -> Sequence[Tuple[str, int]]:
+    buckets = []
+    total_count = 0
+    for index, value in enumerate(metric.point.bucket_counts):
+        total_count += value
+        buckets.append(
+            (
+                f"{metric.point.explicit_bounds[index]}",
+                total_count,
+            )
+        )
+    return buckets
+
+
+class PrometheusMetricReader(MetricReader):
+    """Prometheus metric exporter for OpenTelemetry.
+
+    Args:
+        prefix: single-word application prefix relevant to the domain
+            the metric belongs to.
+    """
+
+    def __init__(self, prefix: str = "") -> None:
+        super().__init__()
+        self._collector = _CustomCollector(self.collect, prefix)
+        core.REGISTRY.register(self._collector)
+
+    def _receive_metrics(self, metrics: Iterable[Metric]) -> None:
+        if metrics is None:
+            return
+        token = attach(set_value(_SUPPRESS_INSTRUMENTATION_KEY, True))
+        try:
+            self._collector.add_metrics_data(metrics)
+        except Exception as e:  # pylint: disable=broad-except,invalid-name
+            _logger.exception("Exception while exporting metrics %s", str(e))
+        detach(token)
+
+    def shutdown(self) -> bool:
+        core.REGISTRY.unregister(self._collector)
+        return True
+
+
+class _CustomCollector:
+    """_CustomCollector represents the Prometheus Collector object
+
+    See more:
+    https://github.com/prometheus/client_python#custom-collectors
+    """
+
+    def __init__(self, callback, prefix: str = ""):
+        self._prefix = prefix
+        self._callback = callback
+        self._metrics_to_export = collections.deque()
+        self._non_letters_digits_underscore_re = re.compile(
+            r"[^\w]", re.UNICODE | re.IGNORECASE
+        )
+
+    def add_metrics_data(self, export_records: Sequence[Metric]) -> None:
+        """Add metrics to Prometheus data"""
+        self._metrics_to_export.append(export_records)
+
+    def collect(self) -> None:
+        """Collect fetches the metrics from OpenTelemetry
+        and delivers them as Prometheus Metrics.
+        Collect is invoked every time a ``prometheus.Gatherer`` is run
+        for example when the HTTP endpoint is invoked by Prometheus.
+        """
+        if self._callback is not None:
+            self._callback()
```

Comment:
wondering if this were to return metric sequence you could simplify this chain of calls `callback` -> `_receive_metrics` -> `add_metrics_data` -> `deque` -> `yield`.

Reply:
It's a good question, i don't know how to simplify this since the callback calls `collect` in the MetricReader which calls `_collect` is set by the meter provider's call to `_set_collect_callback`. That then calls `_receiver_metrics` which right now calls the custom collector to pass the metrics. Open to suggestions though

Reply:
`collect` def on other classes returns (sequence of) metric, so if we update reader to do same it will be consistent and as a result you can get rid of the back and forth method calls here?. it doesn't have to be in this pr though; we can address this in another issue if we agree on this. what do you think @open-telemetry/python-approvers ?

Reply:
Sure, I'll open an issue for it.

Reply:
https://github.com/open-telemetry/opentelemetry-python/issues/2500